collection.c

```diff
@@ -2126,15 +2127,17 @@ void mongo_collection_list_indexes_command(zval *this_ptr, zval *return_value TS
 
 	php_mongo_command_cursor_init_from_document(db->link, cmd_cursor, connection->hash, cursor_env TSRMLS_CC);
 
+	/* TODO: Refactor to use MongoCommandCursor::rewind() once it is extracted
+	 * into its own C function (see PHP-1362) */
+	php_mongocommandcursor_fetch_batch_if_first_is_empty(cmd_cursor TSRMLS_CC);
+	cmd_cursor->started_iterating = 1;
+
 	for (
-		php_mongocommandcursor_load_current_element(cmd_cursor TSRMLS_CC);
-		php_mongocommandcursor_is_valid(cmd_cursor);
-		php_mongocommandcursor_advance(cmd_cursor TSRMLS_CC)
+		iter_status = php_mongocommandcursor_load_current_element(cmd_cursor TSRMLS_CC);
+		iter_status == SUCCESS && php_mongocommandcursor_is_valid(cmd_cursor);
+		iter_status = php_mongocommandcursor_advance(cmd_cursor TSRMLS_CC)
```

Comment:
@derickr: Please give this another look. I found a bug in the previous implementation of this PR, which caused us to miss the first element when we called `php_mongocommandcursor_advance()` before the first loop iteration. This diff sticks to the original for loop, but adds:

```c
php_mongocommandcursor_fetch_batch_if_first_is_empty(cmd_cursor TSRMLS_CC);
cmd_cursor->started_iterating = 1;
```

...which were the missing pieces from MongoCommandCursor::rewind() when it handles the "init from document" case.

Also, I added error checking of the load/advance return values. The previous code ignored these and simply relied on `php_mongocommandcursor_is_valid()`. While that was fine for most cases, it meant we could miss an exception thrown from the advance function, which might lead us to iterate over the same element multiple times. There might also have been a risk of an infinite loop. If advance continually returned `FAILURE` from the `cmd_cursor->dead` conditional, as it would repeatedly throw exceptions and this function would `add_next_index_zval()` a duplicate value, ad infinitum.